src/Npgsql/NpgsqlSchema.cs

```diff
@@ -422,14 +509,61 @@ internal static DataTable GetConstraintColumns(NpgsqlConnection conn, [CanBeNull
             }
         }
 
-        internal static DataTable GetDataSourceInformation()
+        static DataTable GetDataSourceInformation(NpgsqlConnection conn)
         {
-            var ds = new DataSet { Locale = CultureInfo.InvariantCulture };
-            LoadMetaDataXmlResource(ds);
-            return ds.Tables["DataSourceInformation"].Copy();
+            var table = new DataTable("DataSourceInformation");
+            var row = table.Rows.Add();
+
+            table.Columns.Add("CompositeIdentifierSeparatorPattern", typeof(string));
+            // TODO: DefaultCatalog? Was in XML (unfilled) but isn't in docs
+            table.Columns.Add("DataSourceProductName", typeof(string));
+            table.Columns.Add("DataSourceProductVersion", typeof(string));
+            table.Columns.Add("DataSourceProductVersionNormalized", typeof(string));
+            table.Columns.Add("GroupByBehavior", typeof(GroupByBehavior));
+            table.Columns.Add("IdentifierPattern", typeof(string));
+            table.Columns.Add("IdentifierCase", typeof(IdentifierCase));
+            table.Columns.Add("OrderByColumnsInSelect", typeof(bool));
+            table.Columns.Add("ParameterMarkerFormat", typeof(string));
+            table.Columns.Add("ParameterMarkerPattern", typeof(string));
+            table.Columns.Add("ParameterNameMaxLength", typeof(int));
+            table.Columns.Add("QuotedIdentifierPattern", typeof(string));
+            table.Columns.Add("QuotedIdentifierCase", typeof(IdentifierCase));
+            table.Columns.Add("ParameterNamePattern", typeof(string));
+            table.Columns.Add("StatementSeparatorPattern", typeof(string));
+            table.Columns.Add("StringLiteralPattern", typeof(string));
+            table.Columns.Add("SupportedJoinOperators", typeof(SupportedJoinOperators));
+
+            var version = conn.PostgreSqlVersion;
+            var normalizedVersion = $"{version.Major:00}.{version.Minor:00}";
+            if (version.Build >= 0)
+                normalizedVersion += $".{version.Build:00}";
+
+            row["CompositeIdentifierSeparatorPattern"] = @"\.";
+            row["DataSourceProductName"] = "Npgsql";
+            row["DataSourceProductVersion"] = version.ToString();
+            row["DataSourceProductVersionNormalized"] = normalizedVersion;
+            row["GroupByBehavior"] = GroupByBehavior.Unrelated;
+            row["IdentifierPattern"] = @"(^\[\p{Lo}\p{Lu}\p{Ll}_@#][\p{Lo}\p{Lu}\p{Ll}\p{Nd}@$#_]*$)|(^\[[^\]\0]|\]\]+\]$)|(^\""[^\""\0]|\""\""+\""$)";
+            row["IdentifierCase"] = IdentifierCase.Insensitive;
+            row["OrderByColumnsInSelect"] = false;
+            row["ParameterMarkerFormat"] = @"{0}";  // TODO: Not sure
+            row["ParameterMarkerPattern"] = @"@[\p{Lo}\p{Lu}\p{Ll}\p{Lm}_@#][\p{Lo}\p{Lu}\p{Ll}\p{Lm}\p{Nd}\uff3f_@#\$]*(?=\s+|$)";
+            row["ParameterNameMaxLength"] = 128; // TODO: Check
```

Comment:
PostgreSQL doesn't use parameter names, but there is [a limit for identifiers (63 bytes max)](https://www.postgresql.org/docs/current/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS) which could be applied for out parameters. On the Npgsql side we do not enforce the restriction, and this probably may cause problems.

Reply:
Well, as you say PostgreSQL doesn't use parameter names - this is purely an Npgsql feature. We translate textual parameter names to PostgreSQL positional placeholders, so in theory there's no actual limit to parameter names...

For out parameters, if someone attempts to create a function with a long output parameter name, I'm pretty sure PostgreSQL will simply truncate that to 63 like any other identifier. If the user then tries to read that output parameter via the long, original name, they will fail...

Am modifying this to 63 just in case, it's not going to have a huge effect in any case.